File: python/cliptoe.py
#!/usr/bin/env python3

import re
import json
import argparse
import signal
from sys import argv
from csv import DictReader
from time import sleep

version = '1.0.1 [python 3] beta'
author = 'underwood@underwood.network'


def signal_handler(sig, frame):
    print("\33[?25h")
    exit(0)


def version_and_author():
    return "{} version {} - {}".format(argv[0], version, author)


def compounds():
    c = {'C21H30O2': 'Tetrahydrocannabinol',
         'H2SO4': 'Sulphuric Acid',
         'NH3': 'Ammonia',
         'SF6': 'Sulphur Hexafluoride',
         'C7H8': 'Toluene',
         'H2O2': 'Hydrogen Peroxide',
         'C2H2': 'Acetylene',
         'CH3OH': 'Methanol',
         'C7H5N3O6': 'Trinitrotoluene',
         'SO4': 'Sulphur Dioxide',
         'NaHCO3': 'Sodium Bicarbonate',
         'C2H5OH': 'Ethanol',
         'CH4': 'Methane',
         'C8H18O': 'Octanol',
         'C2H6': 'Ethane',
         'C15H31N3O13P2': 'Deoxyribonucleic Acid',
         'C8H10N4O2': 'Caffeine',
         'C8H18': 'Octane',
         'C6H6': 'Benzene',
         'NaCl': 'Sodium Chloride',
         'HCl': 'Hydrogen Chloride',
         'CaCO3': 'Calcium Carbonate',
         'HClO3': 'Chloric Acid',
         'HNO3': 'Nitric Acid',
         'ClH3O': 'Hydrochloric Acid',
         'C6H14O': 'Hexanol',
         'C10H8': 'Naphthalene',
         'CO2': 'Carbon Dioxide',
         'H2O': 'Dihydrogen Monoxide',
         'C20H25N3O': 'Lysergic Acid Diethylamide',
         'C6H14': 'Hexane'}
    return c


def list_compounds(compounds):
    pad = 4
    longest = max(map(len, compounds))
    print('{} {}{}'.format('Formula', ' '*(pad+longest-7), 'Common Name'))
    print('{} {}{}'.format('-------', ' '*(pad+longest-7), '-----------'))
    for c in sorted(compounds):
        print('{} {}{}'.format(c, ' '*(pad+longest-len(c)), compounds[c]))
    exit(0)


def epg(elements):
    epg = {}
    # Populate dictionary of elements by period and group
    for i in range(0, 118):
        p = int(elements[i]['period'])
        g = int(elements[i]['group'])
        epg[p, g] = elements[i]
    return epg


def elements_hash(elements):
    ebs = {}
    # Populate dictionary of elements by period and group
    for i in range(0, 118):
        ebs[elements[i]['symbol']] = elements[i]
    return ebs


def get_ptoe_args():
    parser = argparse.ArgumentParser(
            description='Command-line Interactive Periodic Table of Elements')
    parser.add_argument('-va', '--var', help='Colour variation')
    parser.add_argument('-i', '--ind', help='Colour index')
    parser.add_argument(
            '-k', '--key', help='Show colour key', action='store_true')
    parser.add_argument(
            '-l', '--label', help='Show period and group labels',
            action='store_true')
    parser.add_argument(
            '-m', '--mono', help='Disable colours', action='store_true')
    parser.add_argument(
            '-j', '--json', help='Show JSON output', action='store_true')
    parser.add_argument('-a', '--animate', help='Scroll through colours')
    parser.add_argument(
            '-t', '--table', help='Show table', action='store_true')
    parser.add_argument(
            '-d', '--debug', help='Enable debug output', action='store_true')
    parser.add_argument(
            '-v', '--version', help='Print program version',
            action='store_true')
    parser.add_argument(
            '-c', '--compounds', help='List known compounds',
            action='store_true')
    parser.add_argument(
            'query', help='Query to lookup, can be element or compound',
            nargs='?')

    return parser.parse_args()


def ptoe_table(elements, highlighted, ebs):
    output = "\r\n"
    clear = "\33[m"
    # print('')
    if args.label:
        for group in range(1, 19):
            output += "{:>4}".format(group)
        output += "\r\n"
    for period in range(1, 10):
        if args.label:
            output += f"{period } "
        for group in range(1, 19):
            colour = ''
            invert = ''
            uninvert = ''
            # clear = "\33[m"
            symbol = '   '
            try:
                element = epg(elements)[period, group]
                colour = getcolour(element, args, 0)
                symbol = element['symbol']
                if symbol in highlighted:
                    invert = "\33[7m"
                    uninvert = "\33[27m"
                symbol += (' ' * (3 - len(symbol)))
            except KeyError:
                pass
            if args.mono:
                colour = ''
            output += f'{colour}{invert}{symbol}{uninvert}{clear} '
        output += "\r\n"
    output += "\33[m"

    if args.key:
        output += ptoe_colour_key(ebs)

    return output


def ptoe_colour_key(ebs):
    lines = 0
    o = "\r\n"
    for ex in [
        ['H', "Hydrogen"],
        ['He', "G 18"],
        ['F', "G 17"],
        ['O', "G 16"],
        ['N', "G 15"],
        ['C', "G 14"],
        ['B', "G 13"],
        ['Zn', "G 3-12"],
        ['Be', "G 2"],
        ['Li', "G 1-H"],
        ['La', "Lanthanides"],
        ['Ac', "Actinides"],
      ]:
        if (lines % 4) == 0:
            o += "\r\n  "
        lines += 1
        o += getcolour(ebs[ex[0]], args, 1)
        o += f"{ex[0]} {' ' * (2-(len(str(ex[0]))))}\33[m {ex[1]}"
        # o += (' ' * (2-(len(str(ex[0])))))
        o += (' ' * (14-(len(str(ex[1])))))

    o += "\33[m\r\n"
    return o


def ptoe_parse_input(query):
    # Parse chemical formula, input a string and get an array
    # e.g 'H2O' => $VAR1 = [ 'H', '2' ]; $VAR2 = [ 'O', '1' ];
    elem = []
    formula = []
    pattern = '([A-Z][a-z]?)([0-9]+)?'
    matches = re.findall(pattern, query)
    i = 0
    for m in matches:
        element = matches[i][0]
        many = matches[i][1]
        if not many:
            many = 1
        i += 1
        formula.append([element, int(many)])
        elem.append(element)
    return elem, formula


def getcolour(adetails, args, bg):
    inv = ''
    colour = int(adetails['colour'])
    if args.ind:
        colour += int(args.ind)
    if args.var:
        colour *= int(args.var)
    if bg:
        inv = "\33[7m"
    return f"{inv}\33[38;5;{colour}m"


def ptoe_list_atoms(elements):
    args.gmult = 1
    gmresult = re.search('(^\\d+)', args.query)
    if gmresult:
        args.gmult = gmresult.group(1)

    elem, formula = ptoe_parse_input(args.query)

    if args.debug:
        print(formula)
        exit(0)
    elif args.json:
        print(json.dumps(formula, indent=4))
        exit(0)

    higa = {}
    for e in elem:
        higa[e] = True

    if args.table or args.label:
        print(ptoe_table(elements, higa, ebs))

    atmolcalc(formula)


def get_sumoutput(formula):
    sumoutput = []
    atmol = 0
    totalsum = float(0)

    for einc in formula:
        atom = einc[0]
        many = int(einc[1])
        atmol += many
        try:
            adetails = ebs[atom]
        except KeyError:
            print('%invalid input')
            exit(1)

        sum = float(adetails['amu']) * many
        totalsum += float(sum)
        colour = ''
        if not args.mono:
            colour = getcolour(adetails, args, 0)
        amu = f"{adetails['amu']} amu"
        if many > 1:
            amu = f"{many} * {adetails['amu']} amu = "
            amu += str(float(float(adetails['amu']) * many)) + ' amu'

        sumoutput.append(
            # hatom +
            colour +
            adetails['name'] + "\33[m" +
            (' ' * (12 - len(adetails['name']))) +
            f"(atomic number {ebs[atom]['number']}) " +
            f"[{atom}] {amu}"
        )

        if len(formula) == 1:
            a = adetails
            pg = f"Period {a['period']}, group {a['group']}."
            fact = f"{pg} {a['category']}, {a['phase']}.\r\n\r\n{a['fact']}."
            spc = ' ' * 12
            sumoutput.append(spc + fact + "\n")
    return sumoutput, atmol, totalsum


def atmolcalc(formula):
    sumoutput, atmol, totalsum = get_sumoutput(formula)

    if args.query:
        if len(formula) < 1:
            print('%not a valid element')
            exit(1)
        elif len(formula) > 1:
            args.query = re.sub("(^[^A-Z]+)", "", args.query)
            k = ''

            try:
                if compounds()[args.query]:
                    k = ' (' + compounds()[args.query] + ')'
            except KeyError:
                pass

            print(str(len(formula))+' elements in compound '+str(args.query+k))
            print(str(atmol * args.gmult) + ' atom' +
                  ('s' * ((atmol * float(args.gmult)) > 1)) +
                  f" in {args.gmult} {args.query} molecule"
                  + ('s' * (float(args.gmult) > 1)))
        else:
            args.query = re.sub("(^[^A-Za-z]+)", "", args.query)
            args.query = re.sub("([^A-Za-z]+)$", "", args.query)
            print('Element: ' + args.query)

    for s in sumoutput:
        print(s)

    if args.query:
        if len(formula) > 1:
            print(
                "There are {} grams per mole of {}".format(
                    str(float(totalsum)), args.query)
            )

    return


def ptoe(elements, ebs, args):
    if args.query:
        invalidresult = re.search('([^A-Za-z0-9])', args.query)
        if invalidresult:
            print('%unknown symbol in query: "' + invalidresult.group(1) + '"')
            exit(1)

        # Search for element by name if query is not symbol, number, or formula
        if re.search('[a-z]{2,}', args.query) or re.search(
                                                    '\\d[a-z]', args.query):
            found = False
            for element in ebs:
                e = ebs[element]
                nm = e['name']
                if nm.upper() == args.query.upper():
                    found = True
                    args.query = element
                    break
            if not found:
                print('%unknown input')
                exit(1)

    if args.ind and int(args.ind) < 0:
        args.ind = 0
    if args.var and int(args.var) < 0:
        args.var = 0

    if not (args.query or args.table):
        args.table = True

    ptoe_list_atoms(elements)
    return


args = get_ptoe_args()

if args.version:
    print(version_and_author())
    exit(0)

if args.compounds:
    compounds = compounds()
    list_compounds(compounds)

if args.query is None:
    args.query = ''

with open("../elements.csv") as elements_csv:
    elements = list(DictReader(elements_csv))

ebs = elements_hash(elements)  # elements by symbol
args.elem, formula = ptoe_parse_input(args.query)

ptoe(elements, ebs, args)

if args.animate:
    signal.signal(signal.SIGINT, signal_handler)
    args.ind = int(1)
    print("\33[?25l\33[H\33[0J")
    while True:
        print("\33[H")
        ptoe(elements, ebs, args)
        args.ind += 1
        sleep(float(args.animate))

File: perl/cliptoe.pl
#!/usr/bin/env perl

use strict;

use Data::Dumper;
use JSON::XS;
use Getopt::Long qw(GetOptions);
use Time::HiRes qw(usleep);
use Text::CSV qw( csv );

my $ptoe_version = '1.1.2';

# Based on PTOE.BAS
# ,----------------------------------------------,
# | PTOE.BAS - by underwood, searinox, and devon |
# |          - December 2021                     |
# `----------------------------------------------'
# Command Line Interactive Periodic Table of Elements
# ' https://www.youtube.com/watch?v=wlOdUATGuQY
#
# ' "There is no standard set of colors used to identify element groups or other properties.
# ' Colors are selected based on how well the text shows up against them, but mostly it's a
# ' matter of personal preference." https://www.thoughtco.com/color-on-the-periodic-table-608827

our @ptoe_category_examples;
our @ptoe_colours;

# our $mole = 6.02214076e23;

our $elements = {
    'compounds' => {
        'C21H30O2'      => 'Tetrahydrocannabinol',
        'H2SO4'         => 'Sulfuric Acid',
        'NH3'           => 'Ammonia',
        'SF6'           => 'Sulfur Hexafluoride',
        'C7H8'          => 'Toluene',
        'H2O2'          => 'Hydrogen Peroxide',
        'C2H2'          => 'Acetylene',
        'CH3OH'         => 'Methanol',
        'C7H5N3O6'      => 'Trinitrotoluene',
        'SO4'           => 'Sulfur Dioxide',
        'NaHCO3'        => 'Sodium Bicarbonate',
        'C2H5OH'        => 'Ethanol',
        'CH4'           => 'Methane',
        'C8H18O'        => 'Octanol',
        'C2H6'          => 'Ethane',
        'C15H31N3O13P2' => 'Deoxyribonucleic Acid',
        'C8H10N4O2'     => 'Caffeine',
        'C8H18'         => 'Octane',
        'C6H6'          => 'Benzene',
        'NaCl'          => 'Sodium Chloride',
        'HCl'           => 'Hydrogen Chloride',
        'CaCO3'         => 'Calcium Carbonate',
        'HClO3'         => 'Chloric Acid',
        'HNO3'          => 'Nitric Acid',
        'ClH3O'         => 'Hydrochloric Acid',
        'C6H14O'        => 'Hexanol',
        'C10H8'         => 'Naphthalene',
        'CO2'           => 'Carbon Dioxide',
        'H2O'           => 'Dihydrogen Monoxide',
        'C20H25N3O'     => 'Lysergic Acid Diethylamide',
        'C6H14'         => 'Hexane'
    }
};

my $aoh =  csv(in => "../elements.csv", headers => "auto");

for my $entry ( $aoh )
{
    for my $w ( @$entry )
    {
        $elements->{elementbysymbol}->{ $w->{symbol} } = $w;
    }
}

my $allsymbols =
    'HHeLiBeBCNOFNeNaMgAlSiPSClArKCaScTiVCrMnFeCoNiCuZnGaGeAsSeBrKrRbSrYZrNbMoTc'.
    'RuRhPdAgCdInSnSbTeIXeCsBaLaCePrNdPmSmEuGdTbDyHoErTmYbLuHfTaWReOsIrPtAuHgTl'.
    'PbBiPoAtRnFrRaAcThPaUNpPuAmCmBkCfEsFmMdNoLrRfDbSgBhHsMtDsRgCnNhFlMcLvTsOg';

my $cou;

while ( $allsymbols =~ /([A-Z][a-z]?)/g )
{
    $elements->{elementbynumber}->{ ++$cou } = $1;
}

sub ptoe_help
{
    println( 'Command Line Interactive Periodic Table of Elements' );
    println();
    println( 'Usage:' );
    println( $0
          . ' [atom]         show details for element by atomic symbol or number'
    );
    println( $0 . ' [compound]     list elements in compound [q]' );
    println();
    println( 'Options:' );
    println( $0 . ' --animate=<n>  scroll through colours, speed is <n>' );
    println( $0 . ' --var=<n>      set colour variation' );
    println( $0 . ' --ind=<n>      set colour start index' );
    println( $0 . ' --label        show period and group labels' );
    println( $0 . ' --compounds    show list of example compounds' );
    println( $0 . ' --key          show colour key' );
    println( $0 . ' --table [q]    show table / highlight elements on table' );
    println( $0 . ' --mono         disable colours' );
    exit;
}

sub load_ptoe_colours
{
    @ptoe_category_examples = ptoe_category_examples();
    @ptoe_colours           = ptoe_colours();
    return;
}

sub ptoe
{
    my ( $conn ) = @_;
    my $query = $conn->{arg};
    if ( $query =~ /([^A-Za-z0-9])/ )
    {
        return error( 'unknown symbol in query: "' . $1 . '"' );
    }
    if ( ( $query =~ /[a-z]{2,}/ ) || ( $query =~ /\d[a-z]/ ) )
    {
        # Search for element by name if query is not symbol, number, or formula
        my $found;
        for ( keys %{ $elements->{elementbysymbol} } )
        {
            my $nm = $elements->{elementbysymbol}->{$_}->{name};
            if ( uc( $nm ) eq uc( $query ) )
            {
                $found = 1;
                $query = $_;
                last;
            }
        }
        return error( 'invalid input' ) if ( !$found );
    }
    return ptoe_list_compounds()
      if ( $conn->{options}->{compounds} || $conn->{options}->{list} );

    $conn->{options}->{ind} = 0 if ( $conn->{options}->{ind} < 0 );
    $conn->{options}->{var} = 0 if ( $conn->{options}->{var} < 0 );

    if ( !( $query || $conn->{options}->{table} ) )
    {
        $conn->{options}->{table} = 1;
    }

    ptoe_list_atoms( $query, $conn );
    return;
}

sub ptoe_auto
{
    my ( $conn ) = @_;
    print( "\e[H" );
    print( "\e[0J" );
    print( "\e[?25l" );
    while ( 1 )
    {
        $conn->{options}->{ind}++;
        $conn->{options}->{var}++ if ( $conn->{options}->{ind} > 16 );
        if ( $conn->{options}->{ind} > 1 )
        {
            print( "\e[?25l\e[10A" );
            print( "\e[?25l\e[3A" ) if ( $conn->{options}->{key} );
            print( "\e[?25l\e[1A" ) if ( $conn->{options}->{label} );
        }
        print ptoe_table( $conn ) . "\e[?25h";

        # print $conn->{options}->{ind} . ' ' . $conn->{options}->{var};
        usleep( 500000 / $conn->{options}->{auto} );
        $conn->{options}->{ind} = 1 if ( $conn->{options}->{ind} > 64 );
        $conn->{options}->{var} = 1 if $conn->{options}->{var} > 32;
    }
    return;
}

sub ptoe_parse_input
{
    my ( $query ) = @_;
    # Parse chemical formula, input a string and get an array
    # e.g 'H2O' => $VAR1 = [ 'H', '2' ]; $VAR2 = [ 'O', '1' ];
    my @elem;
    while ( $query =~ /([A-Z][a-z]?)([0-9]+)?/g )
    {
        my $element = $1;
        my $many    = $2;
        $many = '1' if !$many;
        push @elem, [ $element, $many ];
    }
    return @elem;
}

sub ptoe_list_atoms
{
    my ( $query, $conn ) = @_;
    chomp $query;
    my @sumoutput;
    $query =~ /(^\d+)/;
    my $gmult = $1;
    $gmult = 1 if !$gmult;

    my @elem = ptoe_parse_input( $query );
    if ( $conn->{options}->{debug} )
    {
        # Query in Perl style
        println( Dumper( $query ) );
        exit;
    }
    elsif ( $conn->{options}->{json} )
    {
        # Query in pretty JSON style
        println( JSON::XS->new->pretty->encode( \@elem ) );
        exit;
    }

    my $higa;
    $higa->{ $_->[0] } = 1 for ( @elem );
    print ptoe_table( $conn, $higa ) if ( $conn->{options}->{table} );
    my $atmol;
    my $totalsum;
    for my $einc ( @elem )
    {
        my $atom = $einc->[0];
        my $many = $einc->[1];
        $atmol += $many;
        my $adetails = $elements->{elementbysymbol}->{$atom};

        if ( !defined $adetails )
        {
            error( 'invalid input' );
            return;
        }

        if ( defined $adetails && !$adetails->{amu} )
        {
            error( "invalid input: $atom" );
            return;
        }

        my $sum = $adetails->{amu} * $many;
        $totalsum += $sum;
        my $hatom = ptoe_highlight_atom( $atom, 1, 1, $conn );
        my $amu   = "$adetails->{amu} amu";
        $amu =
          "$many * $adetails->{amu} amu = " . $adetails->{amu} * $many . ' amu'
          if ( $many > 1 );
        push @sumoutput,
          (     $hatom
              . ( ' ' x ( 12 - length( $adetails->{name} ) ) )
              . "(atomic number $elements->{elementbysymbol}->{ $atom }->{number}) "
              . '['
              . $atom . '] '
              . $amu
              . "\n" );
        if ( ( scalar @elem ) eq 1 )
        {
            my $pg =
                'Period '
              . $adetails->{period}
              . ', group '
              . $adetails->{group} . '.';
            my $fact =
                $pg . ' '
              . $adetails->{category} . ', '
              . $adetails->{phase} . '. '
              . $adetails->{fact} . '.';
            my $spc = ( ' ' x 12 );

            # Word wrap
            $fact =~ s/(.{54}[^\s]*)\s+/$1\n$spc/g;
            push @sumoutput, ( $spc . $fact . "\n" );
        }
    }
    if ( $query )
    {
        if ( ( scalar @elem ) < 1 )
        {
            return error( 'not a valid element' );
        }
        elsif ( ( scalar @elem ) > 1 )
        {
            $query =~ s/(^[^A-Z]+)//g;
            my $k;
            if ( defined $elements->{compounds}->{$query} )
            {
                $k = ' (' . $elements->{compounds}->{$query} . ')';
            }
            println(
                ( ( scalar @elem ) ) . ' elements in compound ' . $query . $k );
            println($atmol * $gmult . ' atom'
                  . ( 's' x ( ( $atmol * $gmult ) > 1 ) )
                  . " in $gmult $query molecule"
                  . ( 's' x ( $gmult > 1 ) ) );
        }
        else
        {
            $query =~ s/(^[^A-Za-z]+)//g;
            $query =~ s/([^A-Za-z]+)$//g;
            println( 'Element: ' . $query );
        }
    }
    delete $conn->{ptoe};
    print @sumoutput;
    if ( $query )
    {
        if ( scalar @elem > 1 )
        {
            my $s = "There are $totalsum grams per mole of $query molecules";
            println( $s );
        }
    }

    # println();
    return;
}

sub ptoe_table
{
    my ( $conn, $hig ) = @_;

    if ( $conn->{options}->{label} )
    {
        my $line;
        for my $g ( 1 .. 18 )
        {
            $line .= ( "\e[2m" . ( ' ' x ( 4 - length( $g ) ) ) . "$g\e[22m" );
        }
        println( $line );
    }
    for my $p ( 1 .. 9 )
    {
        my $line;
        $line .= ( "\e[2m$p\e[22m " ) if ( $conn->{options}->{label} );
        for my $g ( 1 .. 18 )
        {
            my $atomn = ptoe_epg( $p, $g );
            my $a;
            ( !$atomn ) ? $a = '' : $a = $elements->{elementbynumber}->{$atomn};

            $line .= ( ptoe_highlight_atom( $a, 0, $hig->{$a}, $conn ) . ' ' );
        }
        println( $line );
    }
    println();
    return if ( !$conn->{options}->{key} );
    return if ( $conn->{options}->{mono} );

    my $ck = '';
    for my $gy ( 1 .. 12 )
    {
        my $atomn = $ptoe_category_examples[$gy];
        my $b;
        ( !$atomn ) ? $b = '' : $b = $elements->{elementbynumber}->{$atomn};
        $ck .= "  \e[7m" . ptoe_highlight_atom( $b, 0, 1, $conn ) . ' ';
        my $kw = $ptoe_colours[$gy];
        $ck .= $kw . ( ' ' x ( 13 - length( $kw ) ) );
        $ck .= "\r\n" if !( $gy % 4 );
    }
    chomp $ck;
    println( "\e[0J$ck" );
    println() if ( $conn->{arg} );
    return;
}

sub ptoe_highlight_atom
{
    my ( $asym, $full, $hig, $conn ) = @_;
    my $inv;
    $inv = ( "\e[7m" ) if ( $hig );

    if ( $conn->{options}->{mono} )
    {
        return $elements->{elementbysymbol}->{$asym}->{name} if ( $full );
        return "$inv$asym" . ( ' ' x ( 3 - length( $asym ) ) ) . "\e[m"
          if ( $hig );
        return $asym . ( ' ' x ( 3 - length( $asym ) ) );
    }
    return $asym if ( $conn->{options}->{mono} );
    my $c    = $elements->{elementbysymbol}->{$asym}->{colour};
    my $anam = $elements->{elementbysymbol}->{$asym}->{name};
    my $var  = int( $conn->{options}->{var} );
    my $ind  = int( $conn->{options}->{ind} );
    $c += $ind;
    $var = 1   if ( !defined $conn->{options}->{var} );
    $c *= $var if ( $var );
    $c = $ind  if ( defined $var && !$var );

    return "\e[38;5;$c" . "m$anam\e[m" if ( $full );
    return
        "$inv\e[38;5;$c"
      . "m$asym"
      . ( ' ' x ( 3 - length( $asym ) ) ) . "\e[m";

    # return "\e[38;5;$c"."m$asym\e[m";
}

sub ptoe_epg
{
    # Element by Period and Group
    # Returns atomic number
    my ( $p, $g ) = @_;

    for my $e ( keys %{ $elements->{elementbysymbol} } )
    {
        my $anum   = $elements->{elementbysymbol}->{$e}->{number};
        my $period = $elements->{elementbysymbol}->{$e}->{period};
        my $group  = $elements->{elementbysymbol}->{$e}->{group};
        return $anum if ( $p == $period && $g == $group );
    }
    return;
}

sub ptoe_list_compounds
{
    my ( $conn ) = @_;
    my @out;
    my $longest;
    for my $k ( keys %{ $elements->{compounds} } )
    {
        $longest = length( $k ) if ( length( $k ) > $longest );
    }
    println(
        'Formula' . ( ' ' x ( ( $longest + 2 ) - 7 ) ) . 'Name of compound' );
    println(
        '-------' . ( ' ' x ( ( $longest + 2 ) - 7 ) ) . '----------------' );
    for my $k ( sort keys %{ $elements->{compounds} } )
    {
        println($k
              . ( ' ' x ( ( $longest + 2 ) - ( length( $k ) ) ) )
              . $elements->{compounds}->{$k} );
    }
    return;
}

sub ptoe_category_examples
{
    # Example elements in colour categories
    my @p;
    my $i;
    for (
        'H',  'He', 'F',  'O',  'N',  'C',  'B',
        'Zn', 'Be', 'Li', 'La', 'Ac', 'La', 'Ac'
      )
    {
        $p[ ++$i ] = $elements->{elementbysymbol}->{$_}->{number};
    }
    return @p;
}

sub ptoe_colours
{
    my @c;

    # Colour categories
    my $i;
    $c[ ++$i ] = $_
      for (
        "Hydrogen", "G 18", "G 17", "G 16", "G 15",
        "G 14", "G 13", "G 3-12", "G 2", "G 1-H",
        "Lanthanides", "Actinides"
      );
    return @c;
}

sub version_and_author
{
    return
        "This is $0 version "
      . $ptoe_version
      . " by underwood\@telehack.com";
}

sub ptoe_usage
{
    my ( $conn ) = @_;
    my $u = "usage: $0 \[element|compound\] (case sensitive) "
      . "'$0 --help' for full details";
    return $u;
}

sub println
{
    my ( $s ) = @_;
    print( "$s\n" );
    return;
}

sub error
{
    my ( $s ) = @_;
    print( "\e[31m\%$s\e[m\n" );
    println( ptoe_usage() );
    exit;
}

sub do_ptoe
{
    my $conn;

    my $help;
    my $version;
    my $var;
    my $ind;
    my $key;
    my $label;
    my $mono;
    my $auto;
    my $json;
    my $debug;
    my $table;
    my $comp;

    GetOptions(
        "var=i"     => \$var,        # numeric
        "ind=i"     => \$ind,        # numeric
        "key"       => \$key,        # flag
        "label"     => \$label,      # flag
        "mono"      => \$mono,       # flag
        "json"      => \$json,       # flag
        "animate=i" => \$auto,       # numeric
        "table"     => \$table,      # flag
        "debug"     => \$debug,      # flag
        "help"      => \$help,       # flag
        "version"   => \$version,    # flag
        "compounds" => \$comp,       # flag
    ) or error( 'Error in command line arguments' );

    $conn->{options}->{var}       = $var;
    $conn->{options}->{ind}       = $ind;
    $conn->{options}->{key}       = $key;
    $conn->{options}->{label}     = $label;
    $conn->{options}->{mono}      = $mono;
    $conn->{options}->{auto}      = $auto;
    $conn->{options}->{json}      = $json;
    $conn->{options}->{table}     = $table;
    $conn->{options}->{debug}     = $debug;
    $conn->{options}->{help}      = $help;
    $conn->{options}->{version}   = $version;
    $conn->{options}->{compounds} = $comp;

    if ( $help )
    {
        println( version_and_author() );
        ptoe_help();
        exit;
    }
    if ( $version )
    {
        println( version_and_author() );
        exit;
    }
    if ( $auto )
    {
        return ptoe_auto( $conn );
    }

    my $formula = $ARGV[0];
    chomp $formula;
    if ( !$formula && !$table && !$comp )
    {
        $formula = <STDIN>;
    }
    chomp $formula;
    if ( $formula =~ /^[0-9]+$/ )
    {
        if ( $elements->{elementbynumber}->{$formula} )
        {
            $formula = $elements->{elementbynumber}->{$formula};
        }
        else
        {
            error( 'invalid input' );
            exit;
        }
    }
    $conn->{arg} = $formula;
    ptoe( $conn );
    return;
}

sub repaint_cursor
{
    $SIG{INT} = \&repaint_cursor;
    print( "\e[0J" );
    print( "\e[H" );
    print( "\e[?25h^C\e[0K\n" );
    exit;
}

$SIG{INT} = \&repaint_cursor;

load_ptoe_colours();
do_ptoe();

1;
